ci/scripts/check-bazel-tags.sh

```diff
@@ -6,14 +6,119 @@
 
 set -e
 
-untagged=$(./bazelisk.sh query "rdeps(//..., //hw:verilator) except attr(tags, '[\\[ ](verilator|manual)[,\\]]', //...)")
-if [[ ${untagged} ]]; then # Check that all targets that depend on verilator are tagged
-  echo "Target(s): ${untagged} depend(s) on //hw:verilator, please tag it with verilator or manual";
-  exit 1
-fi
+# The list of bazel tags is represented as a string and checked with a regex
+# https://bazel.build/query/language#attr
+# This function takes a tag(or regex component) and wraps it so attr can query
+# for exact matches.
+exact_regex () {
+  echo "[\\[ ]${1}[,\\]]"
+}
 
-untagged=$(./bazelisk.sh query "rdeps(//..., kind('bitstream_splice', //...)) except attr(tags, '[\\[ ](cw310_rom|cw310_test_rom|vivado|manual)[,\\]]', //...)")
-if [[ ${untagged} ]]; then # Check that all targets that depend on vivado are tagged
-  echo "Target(s): ${untagged} depend(s) on a bitstream_splice but isn't tagged with vivado or manual";
+check_empty () {
+if [[ ${1} ]]; then
+  echo "Error:"
+  echo "$1"|sed 's/^/    /';
+  echo "$2"
   exit 1
 fi
+}
+
+# This check ensures OpenTitan software can be built with a wildcard without
+# waiting for Verilator using --build_tag_filters=-verilator
+untagged=$(./bazelisk.sh query \
+  "rdeps(
+      //...,
+      //hw:verilator
+  )
+  except
+  attr(
+      tags,
+      '$(exact_regex "(verilator|manual)")',
+      //...
+  )" \
+  --output=label_kind)
+check_empty "${untagged}" \
+"Target(s) above depend(s) on //hw:verilator, please tag it with verilator or
+(to prevent matching any wildcards) manual.
+NOTE: test_suites that contain targets with different tags should almost
+universally use the manual tag."
+
+# This check ensures Opentitan software can be built with wildcards in
+# environments that don't have vivado or vivado tools installed by using
+# --build_tag_filters=-vivado.
+untagged=$(./bazelisk.sh query \
+  "rdeps(
+      //...,
+      kind(
+          'bitstream_splice',
+          //...
+      )
+  )
+  except
+  attr(
+      tags,
+      '$(exact_regex "(cw310_rom|cw310_test_rom|vivado|manual)")',
+      //...
+  )" \
+  --output=label_kind)
+check_empty "${untagged}" \
+"Target(s) above depend(s) on a bitstream_splice that isn't cached.
+Please tag it with vivado or (to prevent matching any wildcards) manual.
+NOTE: test_suites that contain targets with different tags should almost
+universally use the manual tag."
```

Comment:
I'm a little confused about this message as well.  How about:

> The targets above depend on a bitstream_splice target, but are not appropriately tagged. Please indicate which kind of bitstream each target depends on by tagging it with one of the following: 'cw310_rom', 'cw310_test_rom', or 'vivado'. Optionally, add the 'manual' tag to exclude the targets from Bazel wildcards.

Reply:
The cw310_rom and cw310_test_rom tags get added by the functest rule, that's not likely to be an appropriate resolution.

Reply:
@dmcardle I don't know what you're asking for here so I'm blocked on resolving it.